### DISCUSSION TITLE
Setting a value to a scalar product using Mathematica List

### DISCUSSION THREAD

**Discussion by YangPhy** [FeynCalc/feyncalc#85] [Q&A] [answered]
Hi all,

Today I was trying to set a value to a scalar product using MMA List but it did now work well. Let me show what was happening:

As we know, in FeynCalc, SPD[p] stands for the scalar product SPD[p,p], so it is understandable that SPD[p,p]=0 if we set SPD[p]=0. However, if one wants to set the values of several scalar products together, using {SPD[p1],SPD[p2],SPD[p3]}={0,0,0}, there would be SPD[p1]=0 but SPD[p1,p1]=p1^2--they are not equal to each other anymore!

Does anyone else also see the same thing? 

Attached is an example in .nb file.

[ScalarProduct.zip](https://github.com/FeynCalc/feyncalc/files/5613704/ScalarProduct.zip)

Best,
Yang Ma

**Comment by vsht** [comment, marked as answer]
Unfortunately, it is not possible to set FeynCalc scalar products via direct list assignments.

The code
```
{SPD[p1]}={0}
```
actually assigns `0` only to `SPD[p1]`, but not to the more fundamental `Pair[Momentum[p1,D],Momentum[p1,D]]`,
so it doesn't work as expected neither for `SPD[p1]` nor for `SPD[p1,p1]`.

The reason for this is a technical limitation within Mathematica: Scalar product assignments in FeynCalc are implemented 
via [UpValues](https://reference.wolfram.com/language/ref/UpValues.html) using [TagSetDelayed](https://reference.wolfram.com/language/ref/TagSetDelayed.html). However, the assignments cannot be arbitrarily complicated, as that would slow down Mathematica's
pattern matcher too much. One can have e.g.
```
SP/:
	Set[SP[a_, b_] , c_]:=
		(ScalarProduct[a,b,Dimension->4,SetDimensions->{4}]=c)
```
but a definition such as
```
SP /: Set[{a__SP}, {b__}] := MapThread[Set, {{a}, {b}}]
```
is already too complex and is not allowed.

Of course, one could redefine `Set` itself, as in 
```
Unprotect[Set];
Set[{a__SP}, {b__}] := MapThread[Set, {{a}, {b}}];
Protect[Set];
```
which would make
```
{SP[p1], SP[p2]} = {x2, y2}
```
work as expected. However, messing with the definitions of protected Mathematica symbols is almost always
considered a very poor programming practice and could potentially introduce numerous unwanted side effects.
So I certainly don't want to implement something like that in FeynCalc.



tl;dr Please do not use direct list assignments as in your example. This won't work. 
A possible workaround for your problem would be to define scalar products via the following construction
```
MapThread[Set, {{SPD[p1], SPD[p2], SPD[p3]}, {0, 0, 0}}]
```
This is perfectly fine and should always work as expected. 

I'll consider adding something like `FCSetScalarProducts[{SP[..],SPD[...],...},{val1,val2,...}]` to the dev version
that implements the above idea.

**Reply by YangPhy** [reply]
Thanks for the reply, and I do look forward to the implementation.

**Reply by vsht** [reply]
You can try out FCSetScalarProducts in the dev version.

**Reply by YangPhy** [reply]
Sure, thanks!
